### PR TITLE
[#514] [KMM Support] Migrate to Hilt to Koin

### LINKED ISSUE #514: [KMM Support] Support using Koin for DI instead of Hilt (Dagger)
## Why

As presented via [this presentation](https://docs.google.com/presentation/d/1faEx-hxFHUZJRQuZgy0zIDSzUBd5x1mxqtC1sGSfMeo/edit?usp=sharing), we need to **support using Koin for DI instead of Hilt (Dagger)** to support the KMM Templates.

- Hilt does not support KMM yet (https://github.com/google/dagger/issues/2349).
- Koin is the recommended DI for KMM.
- All project component’s structure will remain. Only DI configuration changes.
- DI with Koin will kept in a separate branch as a second DI option to switch from the script.

## Who Benefits?

Developers

### PR DESCRIPTION
**Pull request by luongvo**
Closes #514

> [!IMPORTANT]  
> THE BRANCH [develop-koin](https://github.com/nimblehq/android-templates/tree/develop-koin) (DI WITH KOIN) WILL  REMAIN ALONG WITH DEVELOP (DI WITH HILT) TO SWITCH BETWEEN 2 OPTIONS IN THE SCRIPT WHEN GENERATING NEW PROJECTS

## What happened 👀

- Migrate Hilt to Koin https://insert-koin.io/docs/reference/koin-android/start.
- Update tests for Koin.

## Insight 📝

- `hiltViewModel()` 👉 `getViewModel(`.
- No more `@HiltAndroidApp`, `@AndroidEntryPoint`, , `@Inject` usage.
- Define a `initKoin` and call it at `MainApplication`.
- Replace all Hilt modules with Koin modules.
  - `SingletonComponent` or `@Singleton` 👉 `singleOf` or `single { }` / others use `factoryOf`.
  - `@ApplicationContext` 👉 `androidContext()`.
  - `@HiltViewModel` 👉 `viewModelOf`
- Add `stopKoin()` at `tearDown()` in each UITest class to avoid duplication of Koin instance.

## Proof Of Work 📹

The app works properly after migrating.

### CODE REVIEW

File: template-compose/buildSrc/src/main/java/Versions.kt [hunk 1]
```diff
@@ -21,12 +21,11 @@ object Versions {
 
     const val GRADLE = "8.1.2"
 
-    const val HILT = "2.48"
-    const val HILT_NAVIGATION_COMPOSE = "1.0.0"
-
-    const val JAVAX_INJECT = "1"
     const val JUNIT = "4.13.2"
 
+    const val KOIN = "3.3.2"
```
Comment (ryan-conway) [inline template-compose/buildSrc/src/main/java/Versions.kt:26]: Shall we rename this to `KOIN_CORE`?

File: template-compose/buildSrc/src/main/java/Versions.kt [hunk 2]
```diff
@@ -21,12 +21,11 @@ object Versions {
...
+    const val KOIN_ANDROID = "3.3.2"
+    const val KOIN_ANDROIDX_COMPOSE = "3.4.1"
```
Comment (ryan-conway) [inline template-compose/buildSrc/src/main/java/Versions.kt:28]: There is a newer version `3.5.0` available for these three dependencies 🙇🏻

**Review by ryan-conway — COMMENTED** [review]
The rest LGTM 👏🏻

File: template-compose/buildSrc/src/main/java/Versions.kt (hunk 1 above)
Comment (luongvo) [reply template-compose/buildSrc/src/main/java/Versions.kt]: Fixed in https://github.com/nimblehq/android-templates/pull/510/commits/b5f1fd6b92bed5b92b555c5373ba45667acd50da

File: template-compose/buildSrc/src/main/java/Versions.kt (hunk 2 above)
Comment (luongvo) [reply template-compose/buildSrc/src/main/java/Versions.kt]: Fixed in https://github.com/nimblehq/android-templates/pull/510/commits/b5f1fd6b92bed5b92b555c5373ba45667acd50da

**Review by ryan-conway — APPROVED** [review]

**Review by chornerman — APPROVED** [review]

**Review by kaungkhantsoe — APPROVED** [review]

**Pull request merged**